kernel/buddy.c

```diff
@@ -0,0 +1,354 @@
+#include "buddy.h"
+#include "list.h"
+#include "param.h"
+#include "memlayout.h"
+#include "spinlock.h"
+#include "riscv.h"
+#include "defs.h"
+
+// Buddy allocator
+
+static int nsizes;  // the number of entries in bd_sizes array
+
+#define LEAF_SIZE 16          // The smallest block size
+#define MAXSIZE (nsizes - 1)  // Largest index in bd_sizes array
+#define BLK_SIZE(k) ((1L << (k)) * LEAF_SIZE)  // Size of block at size k
+#define HEAP_SIZE BLK_SIZE(MAXSIZE)
+#define NBLK(k) (1 << (MAXSIZE - k))  // Number of block at size k
+#define ROUNDUP(n, sz) \
+  (((((n)-1) / (sz)) + 1) * (sz))  // Round up to the next multiple of sz
+
+typedef struct list Bd_list;
+
+// The allocator has sz_info for each size k. Each sz_info has a free
+// list, an array alloc to keep track which blocks have been
+// allocated, and an split array to to keep track which blocks have
+// been split.  The arrays are of type char (which is 1 byte), but the
+// allocator uses 1 bit per block (thus, one char records the info of
+// 8 blocks).
+struct sz_info {
+  Bd_list free;
+  char *xor_alloc_buddy;
+  char *split;
+};
+typedef struct sz_info Sz_info;
+
+static Sz_info *bd_sizes;
+static void *bd_base;  // start address of memory managed by the buddy allocator
+static struct spinlock lock;
+
+// Return 1 if bit at position index in array is set to 1
+int bit_isset(char *array, int index) {
+  char b = array[index / 8];
+  char m = (1 << (index % 8));
+  return (b & m) == m;
+}
+
+// Set bit at position index in array to 1
+void bit_set(char *array, int index) {
+  char b = array[index / 8];
+  char m = (1 << (index % 8));
+  array[index / 8] = (b | m);
+}
+
+// Clear bit at position index in array
+void bit_clear(char *array, int index) {
+  char b = array[index / 8];
+  char m = (1 << (index % 8));
+  array[index / 8] = (b & ~m);
+}
+
+void buddy_bit_inverse(char *array, int index) {
+  char b = array[index / 16];
+  char m = (1 << ((index % 16) / 2));
+  array[index / 16] = (b ^ m);
+}
+
+int buddy_bit_isset(char *array, int index) {
+ char b = array[index / 16];
```

Comment:
not clear why `array[index / 16]` here. It looks like `bit_set`, but in `bit_set` logic is understandable, here is not